test/datasets/test_iwslt2016.py

```diff
@@ -0,0 +1,83 @@
+import os
+import random
+import string
+from collections import defaultdict
+from unittest.mock import patch
+
+from parameterized import parameterized
+from torchtext.datasets.iwslt2016 import IWSLT2016
+from torchtext.data.datasets_utils import _generate_iwslt_files_for_lang_and_split
+
+from ..common.case_utils import TempDirMixin, zip_equal
+from ..common.torchtext_test_case import TorchtextTestCase
+
+
+def _get_mock_dataset(root_dir, split, src, tgt):
+    """
+    root_dir: directory to the mocked dataset
+    """
+    temp_dataset_dir = os.path.join(root_dir, f"IWSLT2016/2016-01/texts/{src}/{tgt}/{src}-{tgt}/")
+    os.makedirs(temp_dataset_dir, exist_ok=True)
+
+    seed = 1
+    mocked_data = defaultdict(lambda: defaultdict(list))
+    valid_set = "tst2013"
+    test_set = "tst2014"
+
+    files_for_split, _ = _generate_iwslt_files_for_lang_and_split(16, src, tgt, valid_set, test_set)
+    src_file = files_for_split[src][split]
+    tgt_file = files_for_split[tgt][split]
+    for file_name in (src_file, tgt_file):
+        txt_file = os.path.join(temp_dataset_dir, file_name)
+        with open(txt_file, "w") as f:
+            # Get file extension (i.e., the language) without the . prefix (.en -> en)
+            lang = os.path.splitext(file_name)[1][1:]
+            for i in range(5):
+                rand_line = " ".join(
+                    random.choice(string.ascii_letters) for i in range(seed)
+                )
+                dataset_line = f"{rand_line} {rand_line}\n"
+                # append line to correct dataset split
+                mocked_data[split][lang].append(dataset_line)
+                f.write(f'{rand_line} {rand_line}\n')
+                seed += 1
+
+    return list(zip(mocked_data[split][src], mocked_data[split][tgt]))
+
+
+class TestIWSLT2016(TempDirMixin, TorchtextTestCase):
+    root_dir = None
+    patcher = None
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.root_dir = cls.get_base_temp_dir()
+        cls.patcher = patch(
+            "torchdata.datapipes.iter.util.cacheholder.OnDiskCacheHolderIterDataPipe._cache_check_fn", return_value=True
+        )
+        cls.patcher.start()
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.patcher.stop()
+        super().tearDownClass()
+
+    @parameterized.expand([("train", "de", "en"), ("valid", "de", "en")])
+    def test_iwslt2016(self, split, src, tgt):
+        expected_samples = _get_mock_dataset(self.root_dir, split, src, tgt)
+
+        dataset = IWSLT2016(root=self.root_dir, split=split)
+
+        samples = list(dataset)
+
+        for sample, expected_sample in zip_equal(samples, expected_samples):
```

Comment:
NIT: can we organize `sampes` and `expected_samples` similar to what we do in the [SST2 PR](https://github.com/pytorch/text/pull/1542/files#diff-59e6a3a1cb3a2dd765c0ea1110134a23c94307c3787674d5f00ee6027423b36eR81-R83) for consistency

Reply:
> Unfortunately it's not quite as straightforward unless we want to hardcode the language pairs in `setUpClass`. Otherwise there's no good way to parameterize them. This is required because the expected file name for caching is a function of the (src_lang, tgt_lang, split) which is somewhat unique.

Gotcha, I think it's okay to keep your current implementation. I didn't catch the fact that the ordering mattered here.